Source/utils/endian_int.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+namespace devilution {
+
+template <class T>
+T SwapLE(T in)
+{
+	static_assert((sizeof(T) == 1) || (sizeof(T) == 2) ||
+	              (sizeof(T) == 4) || (sizeof(T) == 8),
+	              "SwapLE called for unsupported size");
+	switch (sizeof(T)) {
+	case 1:
+		return in;
+	case 2:
+		return static_cast<T>(SDL_SwapLE16(static_cast<Uint16>(in)));
+	case 4:
+		return static_cast<T>(SDL_SwapLE32(static_cast<Uint32>(in)));
+	case 8:
+		return static_cast<T>(SDL_SwapLE64(static_cast<Uint64>(in)));
+	}
+}
+
+template <class T>
+T SwapBE(T in)
+{
+	static_assert((sizeof(T) == 1) || (sizeof(T) == 2) ||
+	              (sizeof(T) == 4) || (sizeof(T) == 8),
+	              "SwapBE called for unsupported size");
+	switch (sizeof(T)) {
+	case 1:
+		return in;
+	case 2:
+		return static_cast<T>(SDL_SwapBE16(static_cast<Uint16>(in)));
+	case 4:
+		return static_cast<T>(SDL_SwapBE32(static_cast<Uint32>(in)));
+	case 8:
+		return static_cast<T>(SDL_SwapBE64(static_cast<Uint64>(in)));
+	}
+}
+
+template <class T, T (*f)(T)>
+class endian_int {
+	T value;
+
+public:
+	endian_int() :
+		value(0)
+	{
+	}
+
+	endian_int(const endian_int<T, f>& in) :
+		value(in.value)
+	{
+	}
+
+	endian_int(const T& in) :
+		value(f(in))
+	{
+	}
+
+	endian_int<T, f>& operator=(const endian_int<T, f>& in)
+	{
+		value = in.value;
+		return *this;
+	}
+
+	endian_int<T, f>& operator=(const T& in)
+	{
+		value = f(in);
+		return *this;
+	}
+
+	endian_int<T, f>& operator-=(const T& in)
+	{
+		value = f(f(value) - in);
+		return *this;
+	}
+
+	endian_int<T, f>& operator+=(const T& in)
+	{
+		value = f(f(value) + in);
+		return *this;
+	}
+
+	operator T() const
+	{
+		return f(value);
+	}
+};
+
+template<class N, class T, T (*f)(T)>
+N endian_cast(const endian_int<T, f>& in)
+{
+	return static_cast<N>(static_cast<T>(in));
+}
+
+typedef endian_int<Uint16, SwapBE> Uint16_BE;
```

Comment:
code style: Use `using` instead of typedef (`using Uint16_BE = endian_int<Uint16, SwapBE>;`)